### PR TITLE
Improve Aarch64 performance by removing NEON pipeline stalls.

### PR DESCRIPTION
**Pull request by AdamHillier**
This PR re-orders the computation in the accumulation loop of the 8x4 assembly kernel so as to remove pipeline stalls by increasing the distance between data-dependent instructions.

Single-threaded performance on a Raspberry Pi 4B improves by ~8.5%, and on my (Android) phone by ~5.0%.

## What do these changes do?

Execution latency is the number of cycles that must pass before a result is available in a result register, after an instruction is issued. If a result of an instruction is an operand of the next instruction, the processor is forced to wait until the previous result is available (the pipeline is stalled).

In the [Cortex A72 software optimisation guide](https://static.docs.arm.com/uan0016/a/cortex_a72_software_optimization_guide_external.pdf) I found the execution latencies of Neon instructions for that CPU (note: this is the CPU used in the Raspberry Pi 4B). The important ones for us are:

| Instruction | Description                                      | Execution latency |
|:-----------:|--------------------------------------------------|:-----------------:|
|    `ADDP`   | Vector pairwise add                              |         3         |
|    `CNT`    | Vector byte-level popcount                       |         3         |
|    `DUP`    | Duplicate regular register into vector register  |         8         |
|    `EOR`    | Vector exclusive or                              |         3         |
|    `LD1`    | Vector load (1 register, 16 bytes)               |         5         |
|    `LD1`    | Vector load (2 registers, 32 bytes)              |         6         |
|    `LD1`    | Vector load (4 registers, 64 bytes)              |         8         |
|   `UADALP`  | Vector unsigned add and accumulate long pairwise |         4         |

### The problem

There are quite a few places where pipeline stalls occur in the existing assembly kernels. For example, there is a two-cycle stall in the `BMLA` instruction here:
https://github.com/larq/compute-engine/blob/936f5f0273c2e9b119ff317ea6de7f5498dcb9d5/larq_compute_engine/core/bgemm_kernels_arm64.h#L531-L532

This doesn't just happen between consecutive instructions. Here, there is a two-cycle stall at the `smax` instruction because the execution latency of the `dup` instruction is very high (8 cycles).
https://github.com/larq/compute-engine/blob/936f5f0273c2e9b119ff317ea6de7f5498dcb9d5/larq_compute_engine/core/bgemm_kernels_arm64.h#L840-L850

By far the worst place for pipeline stalls is at the end of the `BMLA` macro, when accumulating the byte-level popcounts. There is a two-cycle stall between each of these instructions:
https://github.com/larq/compute-engine/blob/936f5f0273c2e9b119ff317ea6de7f5498dcb9d5/larq_compute_engine/core/bgemm_kernels_arm64.h#L536-L540

### The solution

At the moment, the BMLA macro quite logically performs the `xor`, then the `cnt` (the byte-level pop-count), then accumulates the result and adds it to a destination register. Unfortunately, this makes the sequence of pipeline stalls at the end of the macro unavoidable, because there are no instructions that can be inserted between the data-dependent `addp`/`uadalp` instructions. 

However, in theory while accumulating the `cnt` results we could already be computing the `xor` for the next loop iteration. These `xor` instructions could be scheduled inbetween the data-dependent accumulate instructions of the previous iteration, removing the pipeline stalls (and making the `xor`s effectively 'free').

This PR replaces the `BMLA` macro with a `CNT_ACCUM_XOR` alternative, which assumes that the `xor` of the current iteration has already been computed, then performs the `CNT` and accumulate in parallel with the `xor` of the next iteration. The new macro has *no* pipeline stalls (using the execution latencies from the table above), because at least three other instructions lie between every data-dependent instruction pair.

### Caveat

In order to do this, the new macro needs a lot more registers than the old one did. To free up the necessary register space, I have removed the `RUY_OPT_MAX_STREAMING` part of the existing kernel. However, as discussed in #350 the `RUY_OPT_MAX_STREAMING` part of the kernel was not noticably improving performance, and even if it were this decision is justified by the benchmarks below.

## How Has This Been Tested?

CI. The 'big' kernel tests that don't run on CI have been tested locally (and pass).

## Benchmark Results

I have benchmarked current master (the 'Baseline') and this PR on a Raspberry Pi Model 4B and on my Android phone (a OnePlus 6). I used the LCE benchmark utility with `num_runs=250` and report the average and the standard-deviation.

#### Raspberry Pi 4B

|      Model     | Threads |    Baseline   |       PR      | Relative change |
|----------------|:-------:|:-------------:|:-------------:|:---------------:|
|    QuickNet    |    1    | 35.00 +- 0.06 | 31.98 +- 0.04 |      -8.6%      |
|  QuickNetLarge |    1    | 52.81 +- 0.08 | 48.75 +- 0.14 |      -7.7%      |
|   QuickNetXL   |    1    | 92.49 +- 0.14 | 83.99 +- 0.14 |      -9.2%      |
|    QuickNet    |    4    | 17.48 +- 0.10 | 17.02 +- 0.05 |      -2.6%      |
|  QuickNetLarge |    4    | 25.58 +- 0.09 | 24.39 +- 0.21 |      -4.7%      |
|   QuickNetXL   |    4    | 41.85 +- 0.12 | 39.09 +- 0.11 |      -6.6%      |

#### OnePlus 6

|      Model     | Threads |    Baseline   |       PR      | Relative change |
|----------------|:-------:|:-------------:|:-------------:|:---------------:|
|    QuickNet    |    1    | 18.70 +- 0.21 | 17.83 +- 0.19 |      -4.7%      |
| QuickNet Large |    1    | 27.27 +- 0.22 | 26.02 +- 0.22 |      -4.6%      |
|   QuickNet XL  |    1    | 48.05 +- 0.28 | 45.27 +- 0.35 |      -5.8%      |
|    QuickNet    |    4    | 10.08 +- 0.32 |  9.79 +- 0.23 |      -2.9%      |
| QuickNet Large |    4    | 14.20 +- 0.30 | 13.84 +- 0.26 |      -2.5%      |
|   QuickNet XL  |    4    | 25.46 +- 0.42 | 24.72 +- 0.50 |      -2.9%      |

The performance improvement is greater on the Raspberry Pi (average 8.5%/4.7% performance gain for 1/4 threads) than on the OnePlus 6 (5.0%/2.8%).

## Related issue number

N/A.

### CODE REVIEW

**Comment by AdamHillier** [comment]
@lgeiger could we make a new label for `performance` or something like that? Although this isn't a new feature, it seems weird to describe it as an 'internal' improvement because the performance of LCE is the most important thing about it.

**Comment by arashb** [comment]
This is really really good. I think we should consider a special release on top op v0.3.1 to release this earlier and not wait for v0.4!

**Comment by lgeiger** [comment]
> I think we should consider a special release on top of v0.3.1 to release this earlier and not wait for v0.4!

I'd be happy to push a new release based on v0.3.1, though I am not sure how cleanly this will backport since this PR is built ontop of #351 and #389. Cherry-picking #389 won't be an issue, but #351 might need some manual work.

**Comment by AdamHillier** [comment]
With regards to releasing a new version, I'm happy to go either way. I definitely see the merits of getting a new version out sooner, but also if we save this for v0.4 then there'll be more to talk about when that release comes.

File: larq_compute_engine/core/bgemm_kernels_arm64.h [hunk 1]
```diff
@@ -703,72 +700,31 @@ void BinaryKernelNeonOutOfOrder8x4(
       // destination matrix.
       "1:\n"
 
-      LCE_BMLA_HALF(v24, v8, v0, v1, v2, v3, v4, v5, v6, v7)
-
-#if RUY_OPT_ENABLED(RUY_OPT_MAX_STREAMING)
```
[the remark this answers is not in the document]
Comment (AdamHillier) [reply larq_compute_engine/core/bgemm_kernels_arm64.h:708]: It uses an extra set of LHS and RHS registers, and switches between them. It means you can pre-load data into the next bank while still doing computation with the previous bank. This makes a lot of sense if you're memory bound, but we're compute bound so it doesn't help us.

[the remark this answers is not in the document]
Comment (lgeiger) [reply larq_compute_engine/core/bgemm_kernels_arm64.h:708]: Makes sense, thanks for the explanation!

**Pull request merged**